modo/api.py

```diff
@@ -321,3 +323,33 @@ def enrich_metadata(self):
                     self.update_element(inst_names[ele.name], ele)
                 else:
                     continue
+
+    def stream_cram(
+        self, cram_name: str, region: str = None, output: str = None
+    ):
+        """Slices and streams the requested CRAM file, both local and remote,
+        and either outputs a data stream or writes data to local file"""
+
+        # check requested CRAM exists in MODO
+        path = ""
+        filepaths = list(self.list_files())
+        for filepath in filepaths:
+            if cram_name == str(filepath).split("/")[-1]:
+                path = filepath
+                break
+        if path == "":
+            raise ValueError(f"{cram_name} not fount in {self.path}.")
```

Comment:
Technically, we could request the user to specify the file path, not only filename. This would make life easier for us, but I think this will not be the best user experience. The way I see it, part of the idea of having MODO is to switch from thinking of the separate data files to thinking of a single object, which the data files are a part of. Therefore, it is more "natural" from the user point of view to specify the MODO object and the data file associated to it, without the need of specifying the file path. I would rather keep things like this, but if both of you feel strongly against it, then we can change this.

Reply:
I prefer the path:

1. File names are not neccessary unique, but paths are, e.g. `data/ex/reference.fa` and `data/ex/reference/reference.fa` have the same name, but not the same path. Both path are currently possible in modo.Unless we check for unique names upon adding files this would not be safe. 

2. We return the path in `modo.list_files()`and I think we should be consistent and use one way of referencing the same entity.

Reply:
1. I do indeed think that we need to impose unique file names. Moreover, I think that in real life use, files will tend to be unique and more descriptive (I don;t think "reference.fa" is a good representative of file naming that will be used).

2. That's true, though we could have modo.list_files() that will list the filenames and modo.list_filepaths() that will give the paths.

BTW, one problem with a file path given by the user, is that modo.list_files() gives relative paths, while the user can supply an absolute path, which will need to be handled via string manipulation to trim it to a relative path.

Reply:
Instead of the user providing the filename, how about providing the `name` or `id` of a DataEntity?

e.g. if a user runs `modo show`, there will be a record like:
```yaml
/data/demo1:
  '@type': DataEntity
  data_format: CRAM
  data_path: demo1.cram
  description: Dummy data for tests
  has_reference:
  - reference1
  has_sample:
  - sample1
  name: Demo 1
```

Reply:
```py
cram_id = 'demo1'

# all cram files are inside `/data` in the modo metadata hierarchy
cram_record = self.metadata[f"/data/{cram_id}"]["data_path"]

if cram_record["data_format"] != 'CRAM':
    raise ValueError("not a cram record")

cram_path = cram_record["data_path"]
```

Reply:
Changed stream_cram method to use cram_path instead of cram_name (see commit - https://github.com/sdsc-ordes/modo-api/pull/31/commits/0a1a19183218bf9587148dbf07491ddd71c07427)

`    def stream_cram(
        self, cram_name: str, region: str = None, output: str = None
    ):
        """Slices and streams the requested CRAM file, both local and remote,
        and either outputs a data stream or writes data to local file"""

        # check requested CRAM exists in MODO
        if Path(cram_path) not in self.list_files():
            raise ValueError(f"{cram_path} not found in {self.path}.")

        if self.s3_endpoint:
            # http://domain/s3 + bucket/modo/file.cram --> http://domain/htsget/reads/modo/file.cram
            url = (
                self.htsget_endpoint
                + "/reads/"
                + str(Path(*Path(cram_path).parts[1:]))
            )
            # str(Path(*Path(cram_path).parts[1:])) same as path.split("/", maxsplit=1)[1] but cross-platform
            slice_remote_cram(url, region, output)
        else:
            # assuming user did not change directory, filepath should be the
            # relative path to the file.
            iter = slice_cram(cram_path, region)
            return iter`